lib/src/native/realm_core.dart

```diff
@@ -888,6 +888,36 @@ class _RealmCore {
         "Login failed");
     return completer.future;
   }
+
+  static void void_completion_callback(Pointer<Void> userdata, Pointer<realm_app_error> error) {
+    final Completer<void>? completer = userdata.toObject();
+    if (completer == null) {
+      return;
+    }
+
+    if (error != nullptr) {
+      final message = error.ref.message.cast<Utf8>().toDartString();
+      completer.completeError(RealmException(message));
+      return;
+    }
+
+    completer.complete();
+  }
+
+  Future<void> appEmailPasswordRegisterUser(Application application, String email, String password) {
+    final completer = Completer<void>();
+    using((arena) {
+      _realmLib.invokeGetBool(() => _realmLib.realm_app_email_password_provider_client_register_email(
+            application.handle._pointer,
+            email.toUtf8Ptr(arena),
+            password.toRealmString(arena).ref,
+            Pointer.fromFunction(void_completion_callback),
+            completer.toPersistentHandle(),
```

Comment:
That completer should be released if we use persistent handle. My suggestion would be to model methods like this like:

```dart
Future<T> doSomethinAsync(...) {
  final completer = Completer<T>();
  final completerHandle = completer.toPersistentHandle();
  try {
    _realmLib.do_the_async_action(completerHandle);
    final result = await completer.future;
    _realmLib.release_persistent_handle(completerHandle);
    return result;
  } catch (e) {
    _realmLib.release_persistent_handle(completerHandle);
    rethrow;
  }
}
```

Reply:
Alternatively, we may be able to use a weak handle, as long as we await the future inside the method, thus keeping the completer alive:

```dart
Future<T> doSomethinAsync(...) {
  final completer = Completer<T>();
  _realmLib.do_the_async_action(completer.toWeakHandle());
  return await completer.future;
}
```

Reply:
Brilliant 👍 .. I wish I had thought of that 🥇

Reply:
.. actually I'm not sure completer is guaranteed to survive to the end of the scope .. but it looked good at first

Reply:
The completer is deleted in the current code. The _deletePersistentHandlePtr will do that.

Reply:
@nirinchev ^

Reply:
no the completer will not survive if we await there. The future does not hold onto the completer. But since this is an async method I will change it to await anyways

Reply:
I changed my mind and reverted to non async method

Reply:
I see, I missed the delete userdata function that we're passing.